### PR TITLE
Fix mypy typing for `utilities.auto_restart`

### PR DESCRIPTION
**Pull request by stancld**
Still, WIP and some mypy issues need to be fixed.

## What does this PR do?

Part of #7037. This PR disentangles large #7218. It fixes mypy typing in `utilities.auto_restart.py`

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -400,28 +401,31 @@ def _cycle_to_next_worker_and_reset(dataloader: DataLoader, state_dict: Dict[str
...
+    out: Dict[Union[int, str], Union[Dict[str, int], Optional[int]]] = {}
     if iterator is not None:
-        out.update(_find_current_worker(iterator))
+        for iterator_k, iterator_v in _find_current_worker(iterator).items():
+            out[iterator_k] = iterator_v
 
     if not isinstance(dataloader.dataset, CaptureIterableDataset):
         fast_forward_sampler = _find_fast_forward_samplers(dataloader)
         if fast_forward_sampler is not None:
-            out.update(fast_forward_sampler.state_dict(num_batches_processed=num_batches_processed))
```
Comment (tchaton) [inline pytorch_lightning/utilities/auto_restart.py:420]: Won't update works ?

Comment (stancld) [reply pytorch_lightning/utilities/auto_restart.py:419]: It returns me the following mypy error on my local:
```
pytorch_lightning/utilities/auto_restart.py:421: error: Argument 1 to "update" of "dict" has incompatible type "Dict[int, Dict[str, int]]"; expected "Mapping[Union[int, str], Union[Dict[str, int], int, None]]"  [arg-type]
```
